contracts/src/L1/rollup/ScrollChain.sol

```diff
@@ -189,89 +270,80 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
-        // +---------------------+-------------------+------------------+
-        // ^                     ^                   ^
-        // batchPtr              dataPtr             newBatchPtr (re-use var batchPtr)
-        //
-        // 1. We copy the parent batch header from calldata to memory starting at batchPtr
-        // 2. We store `_chunksLength` number of Keccak hashes starting at `dataPtr`. Each Keccak
-        //    hash corresponds to the data hash of a chunk. So we reserve the memory region from
-        //    `dataPtr` to `dataPtr + _chunkLength * 32` for the chunk data hashes.
-        // 3. The memory starting at `newBatchPtr` is used to store the new batch header and compute
-        //    the batch hash.
-
-        // the variable `batchPtr` will be reused later for the current batch
-        (uint256 batchPtr, bytes32 _parentBatchHash) = _loadBatchHeader(_parentBatchHeader);
-
-        uint256 _batchIndex = BatchHeaderV0Codec.batchIndex(batchPtr);
-        uint256 _totalL1MessagesPoppedOverall = BatchHeaderV0Codec.totalL1MessagePopped(batchPtr);
-        require(committedBatches[_batchIndex] == _parentBatchHash, "incorrect parent batch hash");
-        require(committedBatches[_batchIndex + 1] == 0, "batch already committed");
+        if (_chunks.length == 0) revert ErrorBatchIsEmpty();
 
-        // load `dataPtr` and reserve the memory region for chunk data hashes
-        uint256 dataPtr;
-        assembly {
-            dataPtr := mload(0x40)
-            mstore(0x40, add(dataPtr, mul(_chunksLength, 32)))
+        (, bytes32 _parentBatchHash, uint256 _batchIndex, uint256 _totalL1MessagesPoppedOverall) = _loadBatchHeader(
+            _parentBatchHeader
+        );
+        unchecked {
+            _batchIndex += 1;
         }
+        if (committedBatches[_batchIndex] != 0) revert ErrorBatchIsAlreadyCommitted();
 
-        // compute the data hash for each chunk
+        bytes32 _batchHash;
+        uint256 batchPtr;
+        bytes32 _dataHash;
         uint256 _totalL1MessagesPoppedInBatch;
-        for (uint256 i = 0; i < _chunksLength; i++) {
-            uint256 _totalNumL1MessagesInChunk = _commitChunk(
-                dataPtr,
-                _chunks[i],
-                _totalL1MessagesPoppedInBatch,
+        if (_version == 0) {
+            (_dataHash, _totalL1MessagesPoppedInBatch) = _commitChunksV0(
                 _totalL1MessagesPoppedOverall,
+                _chunks,
                 _skippedL1MessageBitmap
             );
-
-            unchecked {
-                _totalL1MessagesPoppedInBatch += _totalNumL1MessagesInChunk;
-                _totalL1MessagesPoppedOverall += _totalNumL1MessagesInChunk;
-                dataPtr += 32;
+            assembly {
+                batchPtr := mload(0x40)
+                _totalL1MessagesPoppedOverall := add(_totalL1MessagesPoppedOverall, _totalL1MessagesPoppedInBatch)
             }
+            // store entries, the order matters
```

Comment:
Why does the order matter?

Reply:
we use assembly to store variables and some of them are not `32` bytes. If we mess the order, the stored variables won't correct.